userspace/libsinsp/container.cpp

```diff
@@ -578,6 +578,8 @@ bool sinsp_container_manager::resolve_container(sinsp_threadinfo* tinfo, bool qu
 		}
 	}
 
+	m_inspector->m_thread_manager->update_pid_for_vpid(tinfo->m_container_id, tinfo->m_vpid, tinfo->m_pid);
```

Comment:
I haven't experimented, but I'm pretty worried about leaving orphaned items in this table. The lifecycle of the entries in this table seems overall disconnected from the lifecycle of the actual threads they refer to. 

For example, you are updating the table every time `sinsp_container_manager::resolve_container()` is called. But `resolve_container` is called multiple times for the same thread exactly because the cgroups might change earlier in the life of the process, as this comment in the code says:

```
//
// Resync container status after an execve, we need to do it
// because at container startup docker spawn a process with vpid=1
// outside of container cgroup and correct cgroups are
// assigned just before doing execve:
//
// 1. docker-runc calls fork() and created process with vpid=1
// 2. docker-runc changes cgroup hierarchy of it
// 3. vpid=1 execve to the real process the user wants to run inside the container
//
```

In such cases, you're going to be adding two elements in the table, but when the process dies you're going to recycle just one of them, leaving the other leaking? I don't know if there are other corner cases like this, this is the first one that came to mind by reading the code.

Reply:
Adding another lookup table is definitely a concern, but I think it's the best we could think of doing (see below). The container is just used as a way to identify the different pid namespaces. Maybe what I can do is modify the implementation of update_pid_for_vpid to ensure a global pid is only referred to once as the value of some container id + vpid key. Other than that, `delete_pid_for_vpid` is called whenever a thread is itself deleted, and the container key is cleaned up when there are no vpids below it.